pkg/certificate/manager.go

```diff
@@ -1,59 +1,122 @@
 package certificate
 
 import (
+	"context"
+	"fmt"
 	"time"
 
 	"github.com/rs/zerolog/log"
 
 	"github.com/openservicemesh/osm/pkg/announcements"
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/constants"
 	"github.com/openservicemesh/osm/pkg/errcode"
 	"github.com/openservicemesh/osm/pkg/k8s/events"
 	"github.com/openservicemesh/osm/pkg/messaging"
 )
 
 // NewManager creates a new CertManager with the passed CA and CA Private Key
-func NewManager(mrcClient MRCClient, serviceCertValidityDuration time.Duration, msgBroker *messaging.Broker) (*Manager, error) {
-	// TODO(#4502): transition this call to a watch function that knows how to handle multiple MRC and can react to changes.
+func NewManager(ctx context.Context, mrcClient MRCClient, serviceCertValidityDuration time.Duration, msgBroker *messaging.Broker, checkInterval time.Duration) (*Manager, error) {
+	m := &Manager{
+		serviceCertValidityDuration: serviceCertValidityDuration,
+		msgBroker:                   msgBroker,
+	}
+	// We need an MRC in order to get the correct issuers, so handle initial MRC List now
 	mrcs, err := mrcClient.List()
 	if err != nil {
 		return nil, err
 	}
 
-	client, ca, clientID, err := mrcClient.GetCertIssuerForMRC(mrcs[0])
+	err = m.init(mrcs, mrcClient)
 	if err != nil {
 		return nil, err
 	}
 
-	c := &issuer{Issuer: client, ID: clientID, CertificateAuthority: ca}
-
-	m := &Manager{
-		// The signingIssuer is responsible for signing all newly issued certificates
-		// The validatingIssuer is the issuer that issued existing certificates.
-		// its underlying cert is still in the validating trust store
-		signingIssuer:               c,
-		validatingIssuer:            c,
-		serviceCertValidityDuration: serviceCertValidityDuration,
-		msgBroker:                   msgBroker,
+	// now that the manager is initialized, start a watch
+	// we wait until the manager is initialized so that there's
+	// no lock contention between the event handler and the initial
+	// state of the manager.
+	mrcEvents, err := mrcClient.Watch(ctx)
+	if err != nil {
+		return nil, err
 	}
-	return m, nil
-}
 
-// Start takes an interval to check if the certificate
-// needs to be rotated
-func (m *Manager) Start(checkInterval time.Duration, stop <-chan struct{}) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, open := <-mrcEvents:
+				if !open {
+					// channel was closed; return
```

Comment:
add a log statement for this case, and for done as well. For the ctx.Done(), probably also check ctx.Err